homeassistant/components/developer_credentials/__init__.py

```diff
@@ -0,0 +1,270 @@
+        developer_credential.client_secret,
+        authorization_server.authorize_url,
+        authorization_server.token_url,
+    )
+    unsub = config_entry_oauth2_flow.async_register_implementation(
+        hass, domain, auth_impl
+    )
+    hass.data[DOMAIN][domain][auth_domain] = unsub
+
+
+def _async_unregister_auth_implementation(
+    hass: HomeAssistant, domain: str, developer_credential: DeveloperCredential
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    unsub = hass.data[DOMAIN][domain].pop(auth_domain)
+    unsub()
+
+
+async def _async_get_authorization_server(
+    hass: HomeAssistant, domain: str
+) -> AuthorizationServer | None:
+    """Return the AuthorizationServer for the integration domain."""
+    if domain not in hass.data[DOMAIN]:
+        return None
+    get_authorization_server = hass.data[DOMAIN][domain][
+        DeveloperCredentialsType.AUTHORIZATION_SERVER.value
+    ]
+    authorization_server = await get_authorization_server(hass)
+    if not authorization_server:
+        return None
+    return authorization_server
+
+
+async def _async_get_developer_credential(
+    hass: HomeAssistant, domain: str
+) -> DeveloperCredential | None:
+    """Return the DeveloperCredential for the integration domain."""
+    if domain not in hass.data[DOMAIN]:
+        return None
+    get_developer_credential = hass.data[DOMAIN][domain][
+        DeveloperCredentialsType.CONFIG_CREDENTIAL.value
+    ]
+    return await get_developer_credential(hass)
+
+
+class DeveloperCredentialsProtocol(Protocol):
+    """Define the format that developer_credentials platforms can have."""
+
+    async def async_get_authorization_server(
+        self, hass: HomeAssistant
+    ) -> AuthorizationServer:
+        """Return authorization server."""
+
+    async def async_get_developer_credential(
+        self, hass: HomeAssistant
+    ) -> DeveloperCredential:
+        """Return a developer credential from configuration.yaml."""
+
+
+async def _register_developer_credentials_platform(
+    hass: HomeAssistant,
+    integration_domain: str,
+    platform: DeveloperCredentialsProtocol,
+):
+    """Register a developer_credentials platform."""
+    get_authorization_server = getattr(platform, "async_get_authorization_server", None)
+    if get_authorization_server is None:
+        return
+    get_config_developer_credential = getattr(
+        platform, "async_get_developer_credential", None
+    )
+    hass.data[DOMAIN][integration_domain] = {
+        DeveloperCredentialsType.AUTHORIZATION_SERVER.value: get_authorization_server,
+        DeveloperCredentialsType.CONFIG_CREDENTIAL.value: get_config_developer_credential,
+    }
+    # Register an authentication implementation for every credential provided
+    if get_config_developer_credential is None:
+        return
+    developer_credential = await _async_get_developer_credential(
```

Comment:
There is a race condition here.

Integrations that want to use this integration need to add it to `dependencies`. That means that it's already set up by the time the integration that needs it is being set up.

The `async_process_integration_platforms` helper listens to `EVENT_COMPONENT_LOADED` to decide to check if the platform exists and register it. 

However the `EVENT_COMPONENT_LOAD` will not fire until the config entries are finished being set up ([source](https://github.com/home-assistant/core/blob/dev/homeassistant/setup.py#L268-L281)).

Reply:
OK, so the order we want is:
- developer credentials setup
- integration setup
- developer credentials platform setup for that integration
- integration config entry

Is that right?

Reply:
With my current limited understanding: this appears to be a fundamental mismatch for a platform, since it needs to run before config entries are created. I am not aware of other use cases that act line this.  Might it make sense to have integrations register in setup rather than using a platform? Or is this something worthwhile to incorporate into a platform feature. Want to get some feedback before acting.

Reply:
Maybe we can make the `config_entry_oauth2_flow.py` helper aware of the application_credentials integration. When ie Nest calls `config_entry_oauth2_flow.async_get_implementations(hass, "nest")` it will try to load the `application_credential` platform for Nest if `application_credential` integration is set up?

Reply:
Ah looks like `config_entry_oauth2_flow.async_add_implementation_provider` may be able to assist with that. I'll have a closer look.

Reply:
OK, i changed this to load via the implementation provider, with some updates to that API. This is nice because there is no logic for registering and de-registering, since everything is loaded as needed from storage.

I also added a first stab at import support since that came up in the docs PR.